### ISSUE TITLE
Moving 4 txp-related items of Scott Fenton to the main set.mm

### ISSUE THREAD

**Issue by mazsa** [metamath/set.mm#2469]
Namely ctxp , df-txp , txprel and txpss3v .
Related relevant conversation is https://github.com/metamath/set.mm/pull/2435 , summary in chronological order:
@sctfn is not against of moving my theorems (impl. his 4 theorems above) to the main: "Personally, I have no issues moving these theorems to main. They're a bit esoteric, but they're perfectly valid."
@jkingdon  (on the 4 theorems above) : "I don't have an especially strong reaction one way or the other although perhaps I would if I knew this material better." 
@avekens : "This may be useful in general - to be discussed, if df-txp (together with all/some related theorems) should be moved to main. In my opinion, nothing speaks against it." and "A Github issue should be opened for a discussion about moving it to main later."
@benjub : "df-txp: I had never heard of the "tail cross"... The "variable free" definitions it permits are interesting, but I'm not sure the range of application justifies its moving to Main." 
(+ I found them usable and very useful in 62 of my theorems not yet submitted to Github.)

**Comment by avekens** [comment]
I searched for "tail Cartesian product" and "tail cross" in Google, but I did not find anything except the metamath pages with Scott's definition and theorems. Is there any poublic reference (book, website) available? First I want to see how this construct is used by others before I can decide if it is useful to be provided in main set.mm.

**Comment by mazsa** [comment]
@avekens I don't know the motivations of @sctfn or about prior use, but this is how I used it:

The first part of my main results is the reproduction of the conventional meaning of partition (e.g. partition ` A ` of  ` X `, [Halmos] p. 28: "A partition of ` X ` is a disjoint collection ` A ` of non-empty subsets of ` X ` whose union is ` X ` ", or Definition 35, [Suppes] p. 83., cf. https://oeis.org/A000110 ) :
mpet3 $p |- ( ( ElDisj A /\ -. (/) e. A ) <-> ( EqvRel ~ A /\ ( U. A /. ~ A ) = A ) )
mpet2 $p |- ( ( `' _E |` A ) Part A <-> ,~ ( `' _E |` A ) ErALT A )
mpet $p |- ( MembPart A <-> MembEr A )
where
df-coels $a |- ~ A = ,~ ( `' _E |` A ) and 
brcoels $p |- ( ( B e. V /\ C e. W ) -> ( B ~ A C <-> E. u e. A ( B e. u /\ C e. u ) ) )

The second part of my main results is not more general than the above old math result (because you cannot set ` R ` in this theorem in a way that you would get ~ mpet2 ), i.e., this is not the hypothetical General Partition Equivalence Theorem gpet |- ( R Part A <-> ,~ R ErALT A ) , but it has a more general part than ~ mpet2 : ` R `, which is good enough for me:
pet2 $p |- ( ( Disj ( R (x) ( `' _E |` A ) ) /\ ( dom ( R (x) ( `' _E |` A ) ) /. ( R (x) ( `' _E |` A ) ) ) = A ) <-> 
  ( EqvRel ,~ ( R (x) ( `' _E |` A ) ) /\ ( dom ,~ ( R (x) ( `' _E |` A ) ) /.  ,~ ( R (x) ( `' _E |` A ) ) ) = A ) ) $=
pet $p |- ( ( R (x) ( `' _E |` A ) ) Part A <-> ,~ ( R (x) ( `' _E |` A ) ) ErALT A )
where
br1cosstxpcnvepres $p |- ( ( ( B e. V /\ C e. W ) /\ ( D e. X /\ E e. Y ) ) ->
  ( <. B , C >. ,~ ( R (x) ( `' _E |` A ) ) <. D , E >. <-> E. u e. A ( ( C e. u /\ u R B ) /\ ( E e. u /\ u R D ) ) ) )

I guess the second part above is not what we could google out. But perhaps this is the best we can say about the connection between partition and equivalence, at least for now. I think it may justify its inclusion in set.mm .

**Comment by sctfn** [comment]
The reference I took it from is "Elementary Set Theory with a Universal
Set" (online at http://www.logic-center.be/cahiers/Copies/CCL10.pdf), p. 36.

**Comment by mazsa** [comment]
(p. 36 of the pdf, p. 40 of the book)

**Comment by mazsa** [comment]
@avekens @benjub @jkingdon do you think the above reference to be sufficient for containing this in the main set.mm (somewhere around Relations http://us.metamath.org/mpeuni/mmtheorems52.html#mm5110s )? Or should I rather duplicate Scott's items in my Mathbox?

**Comment by benjub** [comment]
I think you should duplicate the theorems you need in your mathbox (in the manner mentioned in the previous discussion).

In the notes mentioned above (Holmes, Elementary Set Theory with a Universal Set), the construction corresponding to df-txp is only used in the two paragraphs following the definition, and not in the remaining 200 pages, and is not given a name, which to me confirms that it does not have a wide enough range of applications.

As for the name "tail cartesian product", I think "tail" is more used in computer science, but in math, we would use "range". So "range cartesian product" or "cartesian product on the second factor" would be better names. As for the symbol, \otimes should be reserved for tensor products. A more appropriate symbol could be $\times_R$ or maybe better: $\times_2$. (In the case of functions f: A \to B and g: A \to C, it is common to write (f, g): A \to B \times C, although we cannot do this in set.mm, and commas are already enough overloaded.)

Let's hear form others in the discussions, and maybe also @digama0.

**Comment by sctfn** [comment]
The thing about the construction only being used in the next two paragraphs
is that it is used to prove a metatheorem that is used extensively
elsewhere in the book. That's kind of like saying that ax-2 doesn't matter
because we have the Deduction Theorem. We don't have that construction in
Metamath, so we've got to use ax-2/txp directly. You can see its use in
nf.mm extensively.

For the name, I'm open to renaming. I just needed to call it something.

**Comment by jkingdon** [comment]
I don't really know enough to have much of an opinion of my own (I started reading the Holmes and got very distracted by NFU in general rather than this topic in particular). Based on the above discussion I suppose we should keep it in mathboxes for now but think about maybe moving it to main a bit down the road (especially if it starts acquiring more usage and/or topics like the notation and other details become more clear).

**Comment by mazsa** [comment]
(FYI @randall-holmes )

**Comment by avekens** [comment]
I agree with @benjub and @jkingdon that we should not move ctxp , df-txp , txprel and txpss3v to main set.mm by now. To move them without the other definitions and theorems of section "Quantifier-free definitions" makes no sense, and to move the whole section would be a too drastic action (where would it fit in with the existing structure?).

**Comment by mazsa** [comment]
@avekens there is a logic in what you say but we can't say at the same time that define everything as a class constant ( http://us.metamath.org/mpeuni/mathbox.html 2. guideline) and that we can't find a place for your results when you do. At least not for long.

**Comment by benjub** [comment]
> The thing about the construction only being used in the next two paragraphs is that it is used to prove a metatheorem that is used extensively elsewhere in the book. That's kind of like saying that ax-2 doesn't matter because we have the Deduction Theorem.

It would indeed be ridiculous to say that df-txp does not matter or is useless.  Fortunately, no one said that.  I just read Chapter 6 of Holmes's notes, and saw how it is used to prove the representation metatheorem. And since a metatheorem cannot be used directly in metamath, one needs to resort to its proof rather than its statement. It therefore probably makes sense to define it in nf.mm (http://us2.metamath.org/nfeuni/df-txp.html; though I wouldn't say its use is extensive).  In set.mm, I maintain that it does not have enough applications or is central enough to have its own definition and syntax in Main, at least for now.

I already wrote above that the variable-free definitions it permits are interesting. It's a natural construction, and in the case of functions that I mentioned above, it is very common to consider (f, g): A \to B \times C, so if a real need arises, we may move it to Main (with symbol, for instance, $\times_2$), but only if/when the need arises.

By the way: the case of ax-2 is different since df-txp is a definition (hence is eliminable and conservative) and introduces new syntax.  New syntax is useful as a shorthand when a given pattern occurs often, but it is also easy to be overwhelmed by too much new syntax.

A more important notion is the cartesian product of maps or relations, from which you can obtain the "range cartesian product" by precomposing it with the diagonal embedding \Delta_A : A \to A \times A.  The diagnoal embedding of A is the "range cartesian product" of the identity of A with itself.  I think this would be a good way to motivate the "range cartesian product", and one may even call it the "generalized diagonal embedding", at least in the case of maps.

Looking at your mathbox, I see that you defined the cartesian product in http://us2.metamath.org/mpeuni/df-pprod.html (in terms of the "range cartesian product") and called it the "parallel product".  Given that set.mm is a collaborative work, I would suggest to try and stick to established terminology and notation, especially when it is standard.

**Comment by jkingdon** [comment]
> defined the cartesian product in http://us2.metamath.org/mpeuni/df-pprod.html (in terms of the "range cartesian product") and called it the "parallel product".

Hmm, the relationship is something like `A e. ( _V X. _V ) /\ B e. ( _V X. _V ) -> pprod ( A , B ) = ( A X. B )` right? Combined with http://us.metamath.org/mpeuni/brel.html the pprod theorems do mean that saying something like `X pprod ( A , B ) Y` does have implications that ` X ( A X. B ) Y` does not.

I mean, as with tail product I don't feel like I know this, or its applications, well enough to have much of an opinion about it, but I don't think I'd go as far as saying this is just the same as cartesian product.

**Comment by benjub** [comment]
> Hmm, the relationship is something like `A e. ( _V X. _V ) /\ B e. ( _V X. _V ) -> pprod ( A , B ) = ( A X. B )` right?

No.  What is called the cartesian product of the binary relations A and B is the LHS of that equality, pprod (A , B). It is different from the RHS, the cartesian product of the sets A and B, which may obviously be larger.  It may seem unfortunate, but it is not, and the naming is classical.  The thing is that if A and B are binary relations, then the way they are encoded in set theory should be irrelevant (hence ` ( A X. B ) ` is meaningless).  For instance, I could choose to encode a binary relation A from X to Y as the triple (X, {(x,y) \in X \times Y | xAy }, Y), and this change in the choice of a concrete encoding should have no influence on the meaning of "the cartesian product of the binary relations A and B".

I'm not making this stuff up. It permits to see the cartesian product as a bifunctor in Set, SetRel or similar categories.

If pprod was moved to Main, it should be called the "cartesian product of binary relations" and could have symbol ${}_1\times_2$ (maybe that explains why I chose $\times_2$ for the "range cartesian product" above, which can be seen as the universal map into the cartesian product).

**Comment by jkingdon** [comment]
> What is called the cartesian product of the binary relations A and B

Ah, I see. It is the "of binary relations part" which I didn't understand.

**Comment by sctfn** [comment]
pprod is a little more subtle than that. 


< x , y > pprod ( A , B ) < z , w > <-> ( x A z /\ y B w )

It’s not Cartesian product as I’ve ever heard it.

**Comment by jkingdon** [comment]
> < x , y > pprod ( A , B ) < z , w > <-> ( x A z /\ y B w )

Ah, thanks. I don't know about anyone else on this thread but I was jumping to conclusions based on quickly skimming things. The distinction can be illustrated by comparing the above (which is in set.mm as http://us.metamath.org/mpeuni/brpprod.html ) with the following study:

```
  ${
    $( A study to show how ~ df-pprod might appear slightly similar to ~ df-xp
       but is not.  Compare with ~ brpprod and note ` X A Y ` versus ` X A Z `
       and likewise for ` B ` .  (Contributed by Jim Kingdon, 5-Feb-2022.) $)
    jk-brxp $p |- ( <. X , Y >. ( A X. B ) <. Z , W >.
        <-> ( X A Y /\ Z B W ) ) $=
      ( cop cxp wbr wcel wa brxp df-br anbi12i bitr4i ) DEGZFCGZABHIPAJZQBJZKDE
      AIZFCBIZKPQABLTRUASDEAMFCBMNO $.
  $}
```

(P.S. I know pretty much nothing about the terminology so I have little to say on that front).

**Comment by benjub** [comment]
As I wrote in my two previous messages, pprod **is** the cartesian product of binary relations. I thought I explained the possible confusion above, and gave justification in category theoretic terms, but maybe it's better explained somewhere else:
https://en.wikipedia.org/wiki/Cartesian_product#Cartesian_product_of_functions
https://en.wikipedia.org/wiki/Product_(category_theory)#Discussion
MacLane, Categories for the Working Mathematician, 2nd ed, page 2

**Edit**: two other references, even more explicit:
https://en.wikipedia.org/wiki/Direct_product#Direct_product_of_binary_relations
http://nlab-pages.s3.us-east-2.amazonaws.com/nlab/show/cartesian+product#as_a_functor

**Comment by mazsa** [comment]
@benjub I like the name range Cartesian product = range cross product = range cross = rxp , and @sctfn seems to be not against it either. I don’t like \times_2 which would implicitly imply a \times_R , at least not until we do not define xp as a restriction of an R dimension concept to 2 dimensions as well. As a default I’ll apply say ( A X.. B ) in my Mathbox with Scott’s original htmldefs etc.
htmldef "X.." as
    " <IMG SRC='otimes.gif' WIDTH=13 HEIGHT=19 ALT=' X..' TITLE='X..'> ";
  althtmldef "X.." as " &#x2297; ";
  latexdef "X.." as "\otimes";
Then:
  df-rxp $a |- ( A X.. B ) = ( ( `' ( 1st |` ( _V X. _V ) ) o. A ) i^i ( `' ( 2nd |` ( _V X. _V ) ) o. B ) ) $.
Or, I guess:
 dfrxp2 $p |- ( A X.. B ) = `' { <. <. y , z >. , x >. | ( x A y /\ x B z ) } $. 
 
@avekens I’m not sure we should treat definitions without “moving parts” (0-ary definitions like df-rels below) and/or quantifier-free definitions (like df-txp/df-rxp df-pprod… ) as a block. I think a better approach would be to move the primary version of a definition separately to the place where you find its secondary counterpart (the corresponding predicate) now in set.mm . This means no need for radical structural changes. For example:
 $( Define the relations class. Proper class relations (like ` _I ` , cf. 
     ~ reli ) are not elements of it. The element of this class and the 
     relation predicate are the same when ` R ` is a set (cf.) ~ elrelsrel ). 
     (Contributed by Peter Mazsa, 13-Jun-2018.) $)
  df-rels $a |- Rels = ~P ( _V X. _V ) $.
  $( Define the relation predicate.  Definition 6.4(1) of [TakeutiZaring]
       p. 23.  For alternate definitions, see ~ dfrel2 and ~ dfrel3 .
       (Contributed by NM, 1-Aug-1994.) $)
    df-rel $a |- ( Rel A <-> A C_ ( _V X. _V ) ) $.
  $( The element of the relations class ( ~ df-rels ) and the relation 
     predicate are the same when ` R ` is a set.
     (Contributed by Peter Mazsa, 24-Nov-2018.) $)
  elrelsrel $p |- ( R e. V -> ( R e. Rels <-> Rel R ) ) $=
    ( wcel crels cvv cxp wss wrel elrels2 df-rel syl6bbr ) ABCADCAEEF
    GAHABIAJK $.
If you accept this approach that means that we could handle txp/rxp separately e.g. from pprod as well.

**Comment by mazsa** [comment]
I close this issue now. The status of pprod is unclear for me and for now, @benoit or @scott should open a new issue as soon as one of them have something to say about it or about a generalization of xp and/or xrn and/or pprod (cf. the conversation at the merged PR above) with a reference to this discussion.
